### DISCUSSION TITLE
Binding Aquasuite Octo device on Linux

### DISCUSSION THREAD

**Discussion by gonsaaa** [liquidctl/liquidctl#570] [Q&A] [answered]
Hello,

I have an Aquasuite Octo device which is in the list of supported devices.
When I initialize liquidctl it doesn't detect any devices so I followed the troubleshooting.
It says to bind the device do the kernel.
First I do a dmesg and I found the device:
`[    8.859941] aquacomputer_d5next 0003:0C70:F011.0008: hidraw7: USB HID v1.11 Device [Aqua Computer GmbH & Co. KG OCTO] on usb-0000:2a:00.3-6.1/input1`

I also use the command `lsusb -vt` to show the devices.

```
:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/6p, 480M
    ID 1d6b:0002 Linux Foundation 2.0 root hub
    |__ Port 5: Dev 2, If 0, Class=Human Interface Device, Driver=usbhid, 12M
        ID 1462:7c84 Micro Star International 
    |__ Port 6: Dev 3, If 0, Class=Hub, Driver=hub/4p, 480M
        ID 05e3:0608 Genesys Logic, Inc. Hub
        |__ Port 1: Dev 4, If 0, Class=Vendor Specific Class, Driver=, 12M
            ID 0c70:f011 MCT Elektronikladen 
        |__ Port 1: Dev 4, If 1, Class=Human Interface Device, Driver=usbhid, 12M
            ID 0c70:f011 MCT Elektronikladen
```
It's the last one `ID 0c70:f011`

Now trying to bind this with the command 
`echo '<bus>-<port>:1.0' | sudo tee /sys/bus/usb/drivers/usbhid/bind`

`echo '03-1-6:1.0' | sudo tee /sys/bus/usb/drivers/usbhid/bind `
just says no such device.

Do I have to unbind it first?
Am I inserting the command wrong?

Could someone help me?

**Comment by aleksamagicka** [comment]
Hi,

What's your liquidctl version, as well as your kernel version? Here it's supported since v1.11.0, and in-kernel it gained sensor support since 5.19 (though liquidctl still supports it through direct access on earlier kernels).

Thanks,
Aleksa

**Reply by gonsaaa** [reply]
Thank you for your answer.
I found out I was using 1.5.1 version. Managed to install v1.12.1 and now it detects!
Now, only one fan detects rpm. I do have 6 fans and a pump.
I does say a warning:
`WARNING: some temp sensors cannot be read from aquacomputer_d5next kernel driver`

```
gonsa@whitepop:~$ liquidctl initialize all
Aquacomputer Octo (experimental)
├── Firmware version           1019  
└── Serial number       32098-63430  

gonsa@whitepop:~$ liquidctl status
WARNING: some temp sensors cannot be read from aquacomputer_d5next kernel driver
Aquacomputer Octo (experimental)
├── Sensor 1          34.9  °C
├── Sensor 2          30.5  °C
├── Fan 1 speed          0  rpm
├── Fan 1 power       0.00  W
├── Fan 1 voltage     0.00  V
├── Fan 1 current     0.00  A
├── Fan 2 speed          0  rpm
├── Fan 2 power       0.00  W
├── Fan 2 voltage     0.00  V
├── Fan 2 current     0.00  A
├── Fan 3 speed          0  rpm
├── Fan 3 power       0.00  W
├── Fan 3 voltage     0.00  V
├── Fan 3 current     0.00  A
├── Fan 4 speed          0  rpm
├── Fan 4 power       0.00  W
├── Fan 4 voltage     0.00  V
├── Fan 4 current     0.00  A
├── Fan 5 speed       1697  rpm
├── Fan 5 power       0.00  W
├── Fan 5 voltage    12.11  V
├── Fan 5 current     0.00  A
├── Fan 6 speed          0  rpm
├── Fan 6 power       0.00  W
├── Fan 6 voltage    12.11  V
├── Fan 6 current     0.00  A
├── Fan 7 speed          0  rpm
├── Fan 7 power       0.00  W
├── Fan 7 voltage    12.11  V
├── Fan 7 current     0.00  A
├── Fan 8 speed          0  rpm
├── Fan 8 power       0.00  W
├── Fan 8 voltage    12.11  V
└── Fan 8 current     0.00  A
```

My goal was to control the fans to be able to automate it when CPU temp goes high.
Any advice?

**Reply by aleksamagicka** [reply]
Please try `liquidctl status --direct-access` for the Octo, that will bypass the kernel driver. I'll investigate the zeroed readings, can you try the `sensors` command and write down your kernel version?

**Reply by aleksamagicka** [reply]
Just tried it on 6.1.11 and liquidctl master, reporting for all fans works as it should.

As a heads up, for setting fan speeds, liquidctl currently expects the pwmX_enable hwmon entries to be present and will give you an error about falling back to direct access - that's harmless.

**Comment by gonsaaa** [comment]
You know what, fan speeds are 0 rpm because they indeed are stopped!
So the readings are correct.
now, could you help me trying to do some commands?

I tried:

```
gonsa@whitepop:~$ liquidctl status --direct-access
WARNING: directly reading the status despite aquacomputer_d5next kernel driver
Aquacomputer Octo (experimental)
├── Sensor 1          36.5  °C
├── Sensor 2          33.1  °C
├── Fan 1 speed          0  rpm
├── Fan 1 power       0.00  W
├── Fan 1 voltage     0.00  V
├── Fan 1 current     0.00  A
├── Fan 2 speed          0  rpm
├── Fan 2 power       0.00  W
├── Fan 2 voltage     0.00  V
├── Fan 2 current     0.00  A
├── Fan 3 speed          0  rpm
├── Fan 3 power       0.00  W
├── Fan 3 voltage     0.00  V
├── Fan 3 current     0.00  A
├── Fan 4 speed          0  rpm
├── Fan 4 power       0.00  W
├── Fan 4 voltage     0.00  V
├── Fan 4 current     0.00  A
├── Fan 5 speed       1687  rpm
├── Fan 5 power       0.00  W
├── Fan 5 voltage    12.10  V
├── Fan 5 current     0.00  A
├── Fan 6 speed          0  rpm
├── Fan 6 power       0.00  W
├── Fan 6 voltage    12.10  V
├── Fan 6 current     0.00  A
├── Fan 7 speed          0  rpm
├── Fan 7 power       0.00  W
├── Fan 7 voltage    12.10  V
├── Fan 7 current     0.00  A
├── Fan 8 speed          0  rpm
├── Fan 8 power       0.00  W
├── Fan 8 voltage    12.10  V
└── Fan 8 current     0.00  A

gonsa@whitepop:~$ liquidctl --match octo set Fan6 speed 30
WARNING: required PWM functionality is not available in aquacomputer_d5next kernel driver, falling back to direct access
```

**Reply by aleksamagicka** [reply]
Unless I'm missing something, I don't see any errors. As I said, the warnings are harmless, fan speed will be set either through the driver (which isn't happening because you don't have the dev version of it inserted to your running kernel) or directly, which is what's happening. Support in liquidctl is currently marked as experimental, after all.

**Reply by gonsaaa** [reply]
Still not able to make it work.
`gonsa@whitepop:~$ liquidctl --match Octo set fan speed 20 20 30 30 40 40 55 80 70 100`
`ERROR: Aquacomputer Octo (experimental): operation not supported by the driver`

**Reply by aleksamagicka** [reply]
As the message says, that's [not yet](https://github.com/liquidctl/liquidctl/blob/main/liquidctl/driver/aquacomputer.py#L412) implemented, neither here nor in the [kernel driver](https://github.com/aleksamagicka/aquacomputer_d5next-hwmon/pull/62). Sorry, I haven't yet gotten around to that.

**Comment by gonsaaa** [comment, marked as answer]
So I managed to control the fans and the pump.
Doing `liquidctl set fan1 speed 100` I can hear the fans spinning to max. And doing the same to fan5 (my pump) makes the pump go full speed. I'm already happy with this achievement.
Now, the only thing left I would like to do is set up the speeds with the cpu temperature, which is outside of the Octo sensors.
I can see all the sensors and temperatures (including the cpu) in the program fancontrol with lm-sensors.
Any suggestion on how to achieve this?

As a note, I'm using Pop!_OS which, sadly, is still using kernel 6.1.11. I know that the driver **aquacomputer-d5next** is supported from 6.2 but somehow I can control them.

**Reply by aleksamagicka** [reply]
You can always control them through liquidctl because it isn't necessarily dependent on the kernel driver - it can work on its own without it, but if its available and has support for specific requested functionality, it takes precedence.

As for CPU temp, that's somewhere in `/sys/class/hwmon/hwmonX/tempY_input`, where X is the number of hwmon device and Y is the number of the temp sensor, which depends on the driver that reads your CPU temp. You'll have to find it and somehow watch over its state and run liquidctl at given points. I can't help with that, but if you venture that way perhaps you'll find this [link](https://joonas.fi/2021/07/stable-device-path-for-linux-hwmon-interfaces/) useful.

**Comment by gonsaaa** [comment]
Just wanted to provide an update. I managed to trigger the fans with the temperature. I know there can be very different ways of achieving this but I found a very basic one.
I created a few scripts to set fans to the desired speed.

```
#!/bin/bash
liquidctl set fan1 speed 60
liquidctl set fan2 speed 60
liquidctl set fan3 speed 60
liquidctl set fan4 speed 60
```
Then using Psensor which as an alert trigger in-built 
https://i.imgur.com/0sM5qEW.png

In this case, when the CPU reaches 80º the alarm triggers the script that I provided and the fans go to that speed.

Once, again, thank you for making this program!

**Reply by aleksamagicka** [reply]
Thanks, but I only wrote the aquacomputer part of it 😄 

Glad that you got that figured out.
